File: app/src/main/java/wangdaye/com/geometricweather/settings/activity/PreviewIconActivity.java
package wangdaye.com.geometricweather.settings.activity;

import android.annotation.SuppressLint;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.appcompat.widget.Toolbar;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

import wangdaye.com.geometricweather.R;
import wangdaye.com.geometricweather.basic.GeoActivity;
import wangdaye.com.geometricweather.basic.model.weather.WeatherCode;
import wangdaye.com.geometricweather.resource.ResourceHelper;
import wangdaye.com.geometricweather.resource.provider.ResourceProvider;
import wangdaye.com.geometricweather.resource.provider.DefaultResourceProvider;
import wangdaye.com.geometricweather.resource.provider.PixelResourcesProvider;
import wangdaye.com.geometricweather.resource.provider.ResourcesProviderFactory;
import wangdaye.com.geometricweather.settings.adapter.WeatherIconAdapter;
import wangdaye.com.geometricweather.settings.dialog.AdaptiveIconDialog;
import wangdaye.com.geometricweather.settings.dialog.AnimatableIconDialog;
import wangdaye.com.geometricweather.settings.dialog.MinimalIconDialog;
import wangdaye.com.geometricweather.utils.DisplayUtils;
import wangdaye.com.geometricweather.utils.helpter.IntentHelper;

public class PreviewIconActivity extends GeoActivity {

    private CoordinatorLayout container;

    private ResourceProvider provider;
    private List<WeatherIconAdapter.Item> itemList;

    public static final String KEY_ICON_PREVIEW_ACTIVITY_PACKAGE_NAME
            = "ICON_PREVIEW_ACTIVITY_PACKAGE_NAME";

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_preview_icon);
        initData();
        initWidget();
    }

    @SuppressLint("MissingSuperCall")
    @Override
    protected void onSaveInstanceState(@NonNull Bundle outState) {
        // do nothing.
    }

    @Override
    public View getSnackbarContainer() {
        return container;
    }

    private void initData() {
        provider = ResourcesProviderFactory.getNewInstance(
                getIntent().getStringExtra(KEY_ICON_PREVIEW_ACTIVITY_PACKAGE_NAME)
        );
        itemList = new ArrayList<>();

        itemList.add(new WeatherIconAdapter.Title(getString(R.string.daytime)));
        itemList.add(new WeatherIcon(provider, WeatherCode.CLEAR, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.PARTLY_CLOUDY, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.CLOUDY, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.WIND, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.RAIN, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.SNOW, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.SLEET, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.HAIL, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.THUNDER, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.THUNDERSTORM, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.FOG, true));
        itemList.add(new WeatherIcon(provider, WeatherCode.HAZE, true));
        itemList.add(new WeatherIconAdapter.Line());

        itemList.add(new WeatherIconAdapter.Title(getString(R.string.nighttime)));
        itemList.add(new WeatherIcon(provider, WeatherCode.CLEAR, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.PARTLY_CLOUDY, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.CLOUDY, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.WIND, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.RAIN, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.SNOW, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.SLEET, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.HAIL, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.THUNDER, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.THUNDERSTORM, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.FOG, false));
        itemList.add(new WeatherIcon(provider, WeatherCode.HAZE, false));
        itemList.add(new WeatherIconAdapter.Line());

        boolean darkMode = DisplayUtils.isDarkMode(this);

        itemList.add(new WeatherIconAdapter.Title("Minimal " + getString(R.string.daytime)));
        itemList.add(new MinimalIcon(provider, WeatherCode.CLEAR, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.PARTLY_CLOUDY, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.CLOUDY, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.WIND, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.RAIN, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.SNOW, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.SLEET, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.HAIL, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.THUNDER, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.THUNDERSTORM, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.FOG, true, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.HAZE, true, darkMode));
        itemList.add(new WeatherIconAdapter.Line());

        itemList.add(new WeatherIconAdapter.Title("Minimal " + getString(R.string.nighttime)));
        itemList.add(new MinimalIcon(provider, WeatherCode.CLEAR, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.PARTLY_CLOUDY, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.CLOUDY, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.WIND, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.RAIN, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.SNOW, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.SLEET, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.HAIL, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.THUNDER, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.THUNDERSTORM, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.FOG, false, darkMode));
        itemList.add(new MinimalIcon(provider, WeatherCode.HAZE, false, darkMode));
        itemList.add(new WeatherIconAdapter.Line());

        itemList.add(new WeatherIconAdapter.Title("Shortcuts " + getString(R.string.daytime)));
        itemList.add(new ShortcutIcon(provider, WeatherCode.CLEAR, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.PARTLY_CLOUDY, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.CLOUDY, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.WIND, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.RAIN, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.SNOW, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.SLEET, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.HAIL, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.THUNDER, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.THUNDERSTORM, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.FOG, true));
        itemList.add(new ShortcutIcon(provider, WeatherCode.HAZE, true));
        itemList.add(new WeatherIconAdapter.Line());

        itemList.add(new WeatherIconAdapter.Title("Shortcuts " + getString(R.string.nighttime)));
        itemList.add(new ShortcutIcon(provider, WeatherCode.CLEAR, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.PARTLY_CLOUDY, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.CLOUDY, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.WIND, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.RAIN, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.SNOW, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.SLEET, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.HAIL, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.THUNDER, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.THUNDERSTORM, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.FOG, false));
        itemList.add(new ShortcutIcon(provider, WeatherCode.HAZE, false));
        itemList.add(new WeatherIconAdapter.Line());

        itemList.add(new WeatherIconAdapter.Title(getString(R.string.sunrise_sunset)));
        itemList.add(new SunIcon(provider));
        itemList.add(new MoonIcon(provider));
    }

    private void initWidget() {
        this.container = findViewById(R.id.activity_preview_icon_container);

        Toolbar toolbar = findViewById(R.id.activity_preview_icon_toolbar);
        toolbar.setTitle(provider.getProviderName());
        toolbar.setNavigationIcon(R.drawable.ic_toolbar_back);
        toolbar.setNavigationOnClickListener(v -> finish());
        toolbar.inflateMenu(R.menu.activity_preview_icon);
        toolbar.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_appStore:
                    if (provider instanceof DefaultResourceProvider
                            || provider instanceof PixelResourcesProvider) {
                        IntentHelper.startAppStoreDetailsActivity(this);
                    } else {
                        IntentHelper.startAppStoreDetailsActivity(this, provider.getPackageName());
                    }
                    break;

                case R.id.action_about:
                    if (provider instanceof DefaultResourceProvider
                            || provider instanceof PixelResourcesProvider) {
                        IntentHelper.startApplicationDetailsActivity(this);
                    } else {
                        IntentHelper.startApplicationDetailsActivity(this, provider.getPackageName());
                    }
                    break;
            }
            return true;
        });

        RecyclerView recyclerView = findViewById(R.id.activity_preview_icon_recyclerView);
        GridLayoutManager manager = new GridLayoutManager(this, 4);
        manager.setSpanSizeLookup(WeatherIconAdapter.getSpanSizeLookup(4, itemList));
        recyclerView.setLayoutManager(manager);
        recyclerView.setAdapter(new WeatherIconAdapter(this, itemList));
    }
}

abstract class BaseWeatherIcon extends WeatherIconAdapter.WeatherIcon {

    protected ResourceProvider provider;
    protected WeatherCode weatherCode;
    protected boolean daytime;

    BaseWeatherIcon(ResourceProvider provider, WeatherCode weatherCode, boolean daytime) {
        this.provider = provider;
        this.weatherCode = weatherCode;
        this.daytime = daytime;
    }
}

class WeatherIcon extends BaseWeatherIcon {

    WeatherIcon(ResourceProvider provider, WeatherCode weatherCode, boolean daytime) {
        super(provider, weatherCode, daytime);
    }

    @Override
    public Drawable getDrawable() {
        return ResourceHelper.getWeatherIcon(provider, weatherCode, daytime);
    }

    @Override
    public void onItemClicked(GeoActivity activity) {
        AnimatableIconDialog dialog = new AnimatableIconDialog();
        dialog.setData(
                weatherCode + "_" + (daytime ? "DAY" : "NIGHT"),
                ResourceHelper.getWeatherIcons(provider, weatherCode, daytime),
                ResourceHelper.getWeatherAnimators(provider, weatherCode, daytime)
        );
        dialog.show(activity.getSupportFragmentManager(), null);
    }
}

class MinimalIcon extends BaseWeatherIcon {

    private boolean darkMode;

    MinimalIcon(ResourceProvider provider, WeatherCode weatherCode, boolean daytime, boolean darkMode) {
        super(provider, weatherCode, daytime);
        this.darkMode = darkMode;
    }

    @Override
    public Drawable getDrawable() {
        return ResourceHelper.getWidgetNotificationIcon(
                provider, weatherCode, daytime, true, !darkMode);
    }

    @Override
    public void onItemClicked(GeoActivity activity) {
        MinimalIconDialog dialog = new MinimalIconDialog();
        dialog.setData(
                weatherCode + "_" + (daytime ? "DAY" : "NIGHT"),
                ResourceHelper.getMinimalXmlIcon(provider, weatherCode, daytime),
                ResourceHelper.getWidgetNotificationIcon(
                        provider, weatherCode, daytime, true, "light"),
                ResourceHelper.getWidgetNotificationIcon(
                        provider, weatherCode, daytime, true, "grey"),
                ResourceHelper.getWidgetNotificationIcon(
                        provider, weatherCode, daytime, true, "dark")
        );
        dialog.show(activity.getSupportFragmentManager(), null);
    }
}

class ShortcutIcon extends BaseWeatherIcon {

    ShortcutIcon(ResourceProvider provider, WeatherCode weatherCode, boolean daytime) {
        super(provider, weatherCode, daytime);
    }

    @Override
    public Drawable getDrawable() {
        return ResourceHelper.getShortcutsIcon(provider, weatherCode, daytime);
    }

    @Override
    public void onItemClicked(GeoActivity activity) {
        AdaptiveIconDialog dialog = new AdaptiveIconDialog();
        dialog.setData(
                weatherCode + "_" + (daytime ? "DAY" : "NIGHT"),
                ResourceHelper.getShortcutsForegroundIcon(provider, weatherCode, daytime),
                new ColorDrawable(Color.TRANSPARENT)
        );
        dialog.show(activity.getSupportFragmentManager(), null);
    }
}

class SunIcon extends WeatherIconAdapter.WeatherIcon {

    protected ResourceProvider provider;
    
    SunIcon(ResourceProvider provider) {
        this.provider = provider;
    }

    @Override
    public Drawable getDrawable() {
        return ResourceHelper.getSunDrawable(provider);
    }

    @Override
    public void onItemClicked(GeoActivity activity) {
        // do nothing.
    }
}

class MoonIcon extends SunIcon {

    MoonIcon(ResourceProvider provider) {
        super(provider);
    }

    @Override
    public Drawable getDrawable() {
        return ResourceHelper.getMoonDrawable(provider);
    }
}

File: app/src/main/kotlin/org/breezyweather/ui/settings/activities/PreviewIconActivity.kt
/*
 * This file is part of Breezy Weather.
 *
 * Breezy Weather is free software: you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation, version 3 of the License.
 *
 * Breezy Weather is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
 * or FITNESS FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public
 * License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with Breezy Weather. If not, see <https://www.gnu.org/licenses/>.
 */

package org.breezyweather.ui.settings.activities

import android.annotation.SuppressLint
import android.graphics.drawable.Drawable
import android.os.Bundle
import android.view.MenuItem
import androidx.core.view.updatePadding
import androidx.recyclerview.widget.GridLayoutManager
import breezyweather.domain.weather.reference.WeatherCode
import org.breezyweather.R
import org.breezyweather.common.activities.BreezyActivity
import org.breezyweather.common.extensions.doOnApplyWindowInsets
import org.breezyweather.common.extensions.getThemeColor
import org.breezyweather.common.extensions.isDarkMode
import org.breezyweather.common.utils.ColorUtils
import org.breezyweather.common.utils.helpers.IntentHelper
import org.breezyweather.databinding.ActivityPreviewIconBinding
import org.breezyweather.ui.settings.adapters.WeatherIconAdapter
import org.breezyweather.ui.settings.dialogs.AdaptiveIconDialog
import org.breezyweather.ui.settings.dialogs.AnimatableIconDialog
import org.breezyweather.ui.settings.dialogs.MinimalIconDialog
import org.breezyweather.ui.theme.resource.ResourceHelper
import org.breezyweather.ui.theme.resource.ResourcesProviderFactory
import org.breezyweather.ui.theme.resource.providers.DefaultResourceProvider
import org.breezyweather.ui.theme.resource.providers.ResourceProvider
import java.util.Locale

class PreviewIconActivity : BreezyActivity() {
    private lateinit var binding: ActivityPreviewIconBinding
    private lateinit var mProvider: ResourceProvider
    private val mItemList = mutableListOf<WeatherIconAdapter.Item>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityPreviewIconBinding.inflate(layoutInflater)
        setContentView(binding.root)
        initData()
        initWidget()
    }

    @SuppressLint("MissingSuperCall")
    override fun onSaveInstanceState(outState: Bundle) {
        // do nothing.
    }

    private fun initData() {
        mProvider = ResourcesProviderFactory.getNewInstance(
            intent.getStringExtra(KEY_ICON_PREVIEW_ACTIVITY_PACKAGE_NAME)
        )

        mItemList.clear()

        mItemList.add(WeatherIconAdapter.Title(getString(R.string.daytime)))
        WeatherCode.entries.forEach {
            mItemList.add(WeatherIcon(mProvider, it, true))
        }
        mItemList.add(WeatherIconAdapter.Line())

        mItemList.add(WeatherIconAdapter.Title(getString(R.string.nighttime)))
        WeatherCode.entries.forEach {
            mItemList.add(WeatherIcon(mProvider, it, false))
        }
        mItemList.add(WeatherIconAdapter.Line())

        val darkMode = this.isDarkMode
        mItemList.add(WeatherIconAdapter.Title("Minimal " + getString(R.string.daytime)))
        WeatherCode.entries.forEach {
            mItemList.add(MinimalIcon(mProvider, it, true, darkMode))
        }
        mItemList.add(WeatherIconAdapter.Line())

        mItemList.add(WeatherIconAdapter.Title("Minimal " + getString(R.string.nighttime)))
        WeatherCode.entries.forEach {
            mItemList.add(MinimalIcon(mProvider, it, false, darkMode))
        }
        mItemList.add(WeatherIconAdapter.Line())

        mItemList.add(WeatherIconAdapter.Title("Shortcuts " + getString(R.string.daytime)))
        WeatherCode.entries.forEach {
            mItemList.add(ShortcutIcon(mProvider, it, true))
        }
        mItemList.add(WeatherIconAdapter.Line())

        mItemList.add(WeatherIconAdapter.Title("Shortcuts " + getString(R.string.nighttime)))
        WeatherCode.entries.forEach {
            mItemList.add(ShortcutIcon(mProvider, it, false))
        }
        mItemList.add(WeatherIconAdapter.Line())

        mItemList.add(WeatherIconAdapter.Title(getString(R.string.ephemeris)))
        mItemList.add(SunIcon(mProvider))
        mItemList.add(MoonIcon(mProvider))
    }

    @SuppressLint("NonConstantResourceId")
    private fun initWidget() {
        binding.activityPreviewIconAppBar.injectDefaultSurfaceTintColor()

        binding.activityPreviewIconToolbar.apply {
            title = mProvider.providerName
            setNavigationOnClickListener { finish() }
            inflateMenu(R.menu.activity_preview_icon)
            setOnMenuItemClickListener { item: MenuItem ->
                when (item.itemId) {
                    R.id.action_about ->
                        if (mProvider is DefaultResourceProvider) {
                            IntentHelper.startApplicationDetailsActivity(this@PreviewIconActivity)
                        } else {
                            IntentHelper.startApplicationDetailsActivity(
                                this@PreviewIconActivity,
                                mProvider.packageName
                            )
                        }
                }
                true
            }
            setBackgroundColor(
                ColorUtils.getWidgetSurfaceColor(
                    6f,
                    getThemeColor(androidx.appcompat.R.attr.colorPrimary),
                    getThemeColor(com.google.android.material.R.attr.colorSurface)
                )
            )
        }

        binding.activityPreviewIconRecyclerView.doOnApplyWindowInsets { view, insets ->
            view.updatePadding(
                left = insets.left,
                right = insets.right,
                bottom = insets.bottom
            )
        }
        val manager = GridLayoutManager(this, 4)
        manager.spanSizeLookup = WeatherIconAdapter.getSpanSizeLookup(4, mItemList)
        binding.activityPreviewIconRecyclerView.apply {
            layoutManager = manager
            adapter = WeatherIconAdapter(this@PreviewIconActivity, mItemList)
        }
    }

    companion object {
        const val KEY_ICON_PREVIEW_ACTIVITY_PACKAGE_NAME = "ICON_PREVIEW_ACTIVITY_PACKAGE_NAME"
    }
}

internal abstract class BaseWeatherIcon(
    protected var provider: ResourceProvider,
    protected var weatherCode: WeatherCode,
    protected var daytime: Boolean,
) : WeatherIconAdapter.WeatherIcon() {
    override val contentDescription: String
        get() {
            val name = weatherCode.name.lowercase(Locale.getDefault()).replace("_", " ")
            return name.substring(0, 1).uppercase(Locale.getDefault()) + name.substring(1)
        }
}

internal class WeatherIcon(
    provider: ResourceProvider,
    weatherCode: WeatherCode,
    daytime: Boolean,
) : BaseWeatherIcon(provider, weatherCode, daytime) {
    override val drawable: Drawable
        get() = ResourceHelper.getWeatherIcon(provider, weatherCode, daytime)

    override fun onItemClicked(activity: BreezyActivity) {
        AnimatableIconDialog.show(activity, weatherCode, daytime, provider)
    }
}

internal class MinimalIcon(
    provider: ResourceProvider,
    weatherCode: WeatherCode,
    daytime: Boolean,
    private val mDarkMode: Boolean,
) : BaseWeatherIcon(provider, weatherCode, daytime) {
    override val drawable: Drawable
        get() = ResourceHelper.getWidgetNotificationIcon(provider, weatherCode, daytime, true, !mDarkMode)

    override fun onItemClicked(activity: BreezyActivity) {
        MinimalIconDialog.show(activity, weatherCode, daytime, provider)
    }
}

internal class ShortcutIcon(
    provider: ResourceProvider,
    weatherCode: WeatherCode,
    daytime: Boolean,
) : BaseWeatherIcon(provider, weatherCode, daytime) {
    override val drawable: Drawable
        get() = ResourceHelper.getShortcutsIcon(provider, weatherCode, daytime)

    override fun onItemClicked(activity: BreezyActivity) {
        AdaptiveIconDialog.show(activity, weatherCode, daytime, provider)
    }
}

internal open class SunIcon(
    protected var provider: ResourceProvider,
) : WeatherIconAdapter.WeatherIcon() {
    override val drawable: Drawable
        get() = ResourceHelper.getSunDrawable(provider)

    override val contentDescription: String
        get() = "Sun"

    override fun onItemClicked(activity: BreezyActivity) {
        // do nothing.
    }
}

internal class MoonIcon(
    provider: ResourceProvider,
) : SunIcon(provider) {
    override val drawable: Drawable
        get() = ResourceHelper.getMoonDrawable(provider)

    override val contentDescription: String
        get() = "Moon"
}
